Cannot build project with PackAsTool=true and GeneratePackageOnBuild=true

**Repro**
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.1</TargetFramework>
    <PackAsTool>true</PackAsTool>
    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
  </PropertyGroup>

</Project>
```
`dotnet build`

**Expected**
Build the project *and* produce a nupkg.

**Actual**
`dotnet build` fails

> C:\Program Files\dotnet\sdk\2.1.300-preview2-008523\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(326,5): error MSB4006: There is a circular dependency in the target dependency graph involving target "_WalkEachTargetPerFramework". [C:\dev\natemcmaster\tmp\tmp.csproj]

--- a/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets
@@ -37,36 +37,43 @@ Copyright (c) .NET Foundation. All rights reserved.
     ============================================================
     -->
 
-  <PropertyGroup Condition="'$(NoBuild)' != 'true'">
-    <_BeforePublishTargets>Build</_BeforePublishTargets>
-  </PropertyGroup>
-
-  <PropertyGroup Condition="'$(NoBuild)' == 'true'">
+  <PropertyGroup>
     <!-- We still need to resolve references even if we are not building during publish. -->
     <!-- BuildOnlySettings are required for RAR to find satellites and dependencies -->
-    <_BeforePublishTargets>
+    <_BeforePublishNoBuildTargets>
       BuildOnlySettings;
       _PreventProjectReferencesFromBuilding;
       ResolveReferences;
       PrepareResourceNames;
       ComputeIntermediateSatelliteAssemblies;
-    </_BeforePublishTargets>
+    </_BeforePublishNoBuildTargets>
+
+    <_CorePublishTargets>
+      PrepareForPublish;
+      ComputeAndCopyFilesToPublishDirectory;
+      GeneratePublishDependencyFile;
+      GeneratePublishRuntimeConfigurationFile;
+    </_CorePublishTargets>
   </PropertyGroup>
+
+  <Target Name="_PublishBuildAlternative"
+          Condition="'$(NoBuild)' != 'true'"
+          DependsOnTargets="Build;$(_CorePublishTargets)" />
+
+  <Target Name="_PublishNoBuildAlternative"
+          Condition="'$(NoBuild)' == 'true'"
+          DependsOnTargets="$(_BeforePublishNoBuildTargets);$(_CorePublishTargets)" />
 
   <Target Name="Publish"
           Condition="$(IsPublishable) == 'true'"
-          DependsOnTargets="$(_BeforePublishTargets);
-                            PrepareForPublish;
-                            ComputeAndCopyFilesToPublishDirectory;
-                            GeneratePublishDependencyFile;
-                            GeneratePublishRuntimeConfigurationFile">
+          DependsOnTargets="_PublishBuildAlternative;_PublishNoBuildAlternative" >
 
     <!-- Ensure there is minimal verbosity output pointing to the publish directory and not just the
          build step's minimal output. Otherwise there is no indication at minimal verbosity of where
          the published assets were copied. -->
     <Message Importance="High" Text="$(MSBuildProjectName) -&gt; $([System.IO.Path]::GetFullPath('$(PublishDir)'))" />
   </Target>
-  
+
   <!-- Don't let project reference resolution build project references in NoBuild case. -->
   <Target Name="_PreventProjectReferencesFromBuilding">
     <PropertyGroup>

--- a/src/Tests/Microsoft.NET.ToolPack.Tests/GivenThatWeWantToPackAToolProjectWithGeneratePackageOnBuild.cs
+++ b/src/Tests/Microsoft.NET.ToolPack.Tests/GivenThatWeWantToPackAToolProjectWithGeneratePackageOnBuild.cs
@@ -1,0 +1,93 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Xml.Linq;
+using FluentAssertions;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.NET.TestFramework;
+using Microsoft.NET.TestFramework.Commands;
+using NuGet.Packaging;
+using Xunit;
+using Xunit.Abstractions;
+using Microsoft.NET.TestFramework.Assertions;
+
+namespace Microsoft.NET.ToolPack.Tests
+{
+    public class GivenThatWeWantToPackAToolProjectWithGeneratePackageOnBuild : SdkTest
+    {
+
+        private const string AppName = "consoledemo";
+
+        public GivenThatWeWantToPackAToolProjectWithGeneratePackageOnBuild(ITestOutputHelper log) : base(log)
+        {}
+
+        private TestAsset SetupAndRestoreTestAsset([CallerMemberName] string callingMethod = "")
+        {
+            TestAsset testAsset = _testAssetsManager
+                .CopyTestAsset("PortableToolWithP2P", callingMethod)
+                .WithSource()
+                .WithProjectChanges((projectPath, project) =>
+                {
+                    if (IsAppProject(projectPath))
+                    {
+                        XNamespace ns = project.Root.Name.Namespace;
+                        XElement propertyGroup = project.Root.Elements(ns + "PropertyGroup").First();
+                        propertyGroup.Add(new XElement(ns + "GeneratePackageOnBuild", "true"));
+                    }
+                })
+                .Restore(Log, "App");
+
+            return testAsset;
+        }
+
+        [Fact]
+        public void It_builds_successfully()
+        {
+            TestAsset testAsset = SetupAndRestoreTestAsset();
+            var appProjectDirectory = Path.Combine(testAsset.TestRoot, "App");
+            var buildCommand = new BuildCommand(Log, appProjectDirectory);
+
+            CommandResult result = buildCommand.Execute();
+
+            result.Should()
+                  .Pass()
+                  .And
+                  .NotHaveStdOutContaining("There is a circular dependency");
+        }
+
+        [Fact]
+        public void It_builds_and_result_contains_dependencies_dll()
+        {
+            TestAsset testAsset = SetupAndRestoreTestAsset();
+            var appProjectDirectory = Path.Combine(testAsset.TestRoot, "App");
+            var buildCommand = new BuildCommand(Log, appProjectDirectory);
+            buildCommand.Execute();
+
+            var packCommand = new PackCommand(Log, appProjectDirectory);
+            // Do not run pack, just use it to get nupkg since it should be run by build.
+            var nugetPackage = packCommand.GetNuGetPackage();
+
+            using(var nupkgReader = new PackageArchiveReader(nugetPackage))
+            {
+                IEnumerable<NuGet.Frameworks.NuGetFramework> supportedFrameworks = nupkgReader.GetSupportedFrameworks();
+                supportedFrameworks.Should().NotBeEmpty();
+
+                foreach (NuGet.Frameworks.NuGetFramework framework in supportedFrameworks)
+                {
+                    var allItems = nupkgReader.GetToolItems().SelectMany(i => i.Items).ToList();
+                    allItems.Should().Contain($"tools/{framework.GetShortFolderName()}/any/Library.dll");
+                }
+            }
+        }
+
+        private bool IsAppProject(string projectPath)
+        {
+            return Path.GetFileNameWithoutExtension(projectPath).Equals(AppName, StringComparison.OrdinalIgnoreCase);
+        }
+    }
+}

